engine/derivative.cc

```diff
@@ -639,6 +640,31 @@ namespace MathDAG
        throw error("frac is not differentiable");
   }
   
+  template <>
+  NodePtr SystemOfEquations::derivative<>
+  (const OperationDAG<OperationType::percent>& expr)
+  {
+    if (expr.arguments[0].empty())
+      return zero;
+    else
+       throw error("percent is not differentiable");
+  }
+  
+//  template <>
+//  NodePtr SystemOfEquations::derivative
+//  (const OperationDAG<OperationType::fact>& expr)
+//  {
+//    if (expr.arguments[0].empty())
+//      return zero;
+//    else
+//      {
+//        //Expr x(expressionCache, expr.arguments[0][0]);
+//        //Expr gamma_p = boost::math::digamma(one+x)*boost::math::tgamma(one+x);
```

Comment:
This is a symbolic differentiator. The only way we can do this is to also add in all the polygamma functions ψ⁽ⁿ⁾(x) = dⁿ/dxⁿψ(x) as well.

Fortunately, boost::math provides implementations of all the polygamma functions.

Reply:
I'm not sure, but I would only have to go up to the trigamma function at most, wouldn't I? Otherwise we would need something like the following to deal with derivatives of order 4 or higher, or rather, to calculate nth order derivatives from n-1th ones:

[http://erikerlandson.github.io/blog/2016/06/15/computing-derivatives-of-the-gamma-function/](
http://erikerlandson.github.io/blog/2016/06/15/computing-derivatives-of-the-gamma-function/)

Reply:
I think you only need to provide the derivative for the polygamma function:
d/dxψⁿ(x) = ψⁿ⁺¹(x)
polygamma will need to be implemented as a derived class from operator<polygamma>, with an order parameter attribute.
Where you have digamma now, use polygamma(0), unless there is any use case for making it a separate operation.

Reply:
Thanks, I've done this. :)